### PR TITLE
Initial commit to add Dilithium NIST PQC winner.

### PR DESCRIPTION
**Pull request by anhu**
Also:

* added HAVE_FALCON guards as needed.
* corrected minor falcon bugs as I found them.
* handling OID sum collision between DILITHIUM_LEVEL5 and DILITHIUM_AES_LEVEL3

### CODE REVIEW

**Comment by anhu** [comment]
Tested with the following commands:
```
examples/server/server -v 4 -l TLS_AES_256_GCM_SHA384 \
    -c ~/tmp/dilithium_aes_level5_entity_cert.pem \
    -k ~/tmp/dilithium_aes_level5_entity_key.pem \
    -A ~/tmp/dilithium_aes_level5_root_cert.pem --pqc P521_KYBER_LEVEL5
```

```
examples/client/client -v 4 -l TLS_AES_256_GCM_SHA384 \
    -c ~/tmp/dilithium_aes_level5_entity_cert.pem \
    -k ~/tmp/dilithium_aes_level5_entity_key.pem \
    -A ~/tmp/dilithium_aes_level5_root_cert.pem --pqc P521_KYBER_LEVEL5
```

...with permutations of SHAKE,AES variants and levels 2,3,5

File: gencertbuf.pl [hunk 1]
```diff
@@ -210,6 +222,23 @@
 
 print OUT_FILE "#endif /* HAVE_PQC */\n\n";
 
+# convert and print dilithium keys
+print OUT_FILE "#ifdef HAVE_PQC\n\n";
```
Comment (dgarske) [inline gencertbuf.pl:226]: Should this also be wrapped in `HAVE_DILITHIUM`? Might consider the same for Falcon above.

File: wolfssl/wolfcrypt/dilithium.h [hunk 2]
```diff
@@ -0,0 +1,142 @@
+/* dilithium.h
+ *
+ * Copyright (C) 2021 wolfSSL Inc.
```
Comment (dgarske) [inline wolfssl/wolfcrypt/dilithium.h:3]: 2022

File: wolfcrypt/src/dilithium.c [hunk 3]
```diff
@@ -0,0 +1,802 @@
+/* dilithium.c
+ *
+ * Copyright (C) 2006-2021 wolfSSL Inc.
```
Comment (dgarske) [inline wolfcrypt/src/dilithium.c:3]: 2022

File: wolfcrypt/src/dilithium.c [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+#ifdef HAVE_CONFIG_H
+    #include <config.h>
+#endif
+
+/* in case user set HAVE_PQC there */
+#include <wolfssl/wolfcrypt/settings.h>
+
+#include <wolfssl/wolfcrypt/asn.h>
+
+#ifdef HAVE_PQC
```
Comment (dgarske) [inline wolfcrypt/src/dilithium.c:33]: `HAVE_DILITHIUM` also?

File: wolfcrypt/benchmark/benchmark.c [hunk 5]
```diff
@@ -7295,7 +7305,137 @@ void bench_falconKeySign(byte level)
     wc_falcon_free(&key);
 }
 #endif /* HAVE_FALCON */
-#endif /* HAVE_PQC && HAVE_LIBOQS */
+
+#ifdef HAVE_DILITHIUM
+void bench_dilithiumKeySign(byte level, byte sym)
```
Comment (dgarske) [inline wolfcrypt/benchmark/benchmark.c:7310]: Can you update the documentation with benchmarks for Dilithium?

**Review by dgarske — CHANGES_REQUESTED** [review]
First pass.

File: wolfcrypt/benchmark/benchmark.c (hunk 5 above)
Comment (anhu) [reply wolfcrypt/benchmark/benchmark.c:7310]: I have not done this in the latest update. 
Stay tuned.

**Comment by anhu** [comment]
Force pushed.  Needed to resolve conflicts.

**Review by dgarske — APPROVED** [review]
Code looks great. I would still like to run some tests before merging.

**Comment by dgarske** [comment]
@anhu can you resolve the conflicts? Thanks!

**Comment by anhu** [comment]
Fixed! :)

**Review by dgarske — APPROVED** [review]
Tests out great. Thanks

**Comment by anhu** [comment]
Thanks!! Please merge once tests are done.

**Pull request merged**